### ISSUE TITLE
HuggingFace speech models not supported

### ISSUE THREAD

**Issue by saeid93** [SeldonIO/MLServer#758]
MLserver HuggingFace runtime cannot work with speech models in the batched mode as the pipeline accepts a list of arrays [(request1), (request2), (request3), (request4), (request5)] which the type of each request is a NumPy array. However, MLServer stacked the NumPy data as an array of arrays of shape (batch_size, input_data) which will result in the following error when sending to the HuggingFace pipeline. It thinks the batched inputs are multi-channel inputs rather than batched single-channel inputs.

```bash
raise ValueError("We expect a single channel audio input for AutomaticSpeechRecognitionPipeline")
ValueError: We expect a single channel audio input for AutomaticSpeechRecognitionPipeline
```

**Comment by adriangonz** [comment]
Hey @saeid93 ,

For this one, isn't this a problem of shapes and expected dimensions? 

Looking at the [implementation for AutomaticSpeechRecognitionPipeline](https://huggingface.co/transformers/v4.8.0/_modules/transformers/pipelines/automatic_speech_recognition.html), it seems it expects `inputs` to have a single dimension. Therefore, wouldn't it be enough to ensure that the individual input's shape is always 1-dimensional?

If I'm not mistaken, that will ensure the batching is done across that single dimension, right?

**Comment by saeid93** [comment]
Hey @adriangonz,

Indeed it is. The code you sent is for before sending them as a batch model (single inputs). However when a batch is sent the inputs to the speech model on the MLServer client side like the following:
```python
    payload = {
        "inputs": [
            {
            "name": "array_inputs",
            "shape": [1, len(input_data)],
            "datatype": "FP32",
            "data": input_data.tolist(),
            "parameters": {
                "content_type": "np"
            }
            }
        ]
    }
```
it is received in the MLServer as concatenated as an np.array of shape `(size_of_the_bach, size_of_input_data)`. This won't go through the HuggingFace model since as you said it expects a one-dimensional array. To send batches to a HugggingFace model instead of sending it as a stacked array we need to send it as a list of single dimension `[(1, size_of_input), (1, size_of_input), (1, size_of_input), ...]`. However, in the current implementation sending the above request will result in the error I mentioned in the first comment as it batches in the stacked np.array which the HuggingFace does not accept. In the pull request, I have added support for such cases. It basically turns the stacked np.array format in the batched version into the list of array inputs that is acceptable by the HugggingFace when the input is a NumPy array.

Please let me know if more information is needed.

**Comment by adriangonz** [comment]
Hey @saeid93 ,

I see, thanks for clarifying that. 

My main concern is that the PR changes the behaviour for every HF model, and I'm not sure whether that's something we want.

Instead, a better solution would be to address this on the adaptive batching code, to ensure that it doesn't stack a new batch dimension on top of the existing shape. As far as I can see, adaptive batching should respect the existing `shape`'s dimension (and if it doesn't, that sounds like a bug).

**Comment by saeid93** [comment]
Hey @adriangonz,
That's a good point. It has a check for the np array format so I think it would be unlikely to change every model
```python
X = kwargs.pop("array_inputs", None)
if X is not None:
     args = [list(X)] + args
```

But it might be violated if some other model in HuggingFace is not following the list of items for batching (which as far as I know it is). 

I think the batching code is just fine as it sends the data as a list the model and the user can translate that data in whatever format they eant. It is the HuggingFace runtime that translates it into a stacked numpy (when the input is not text). I'll be happy to investigate this further. However, I want to know your thoughts on any specific case that might get violated by the existing PR. Because the current format is not supporting speech pipelines.

**Comment by adriangonz** [comment]
Well, I think the stacking occurs within the inputs themselves. If I understand right (please correct me if I'm wrong), you are sending them as the snippet below, right?

```json
{
	inputs: [
        { "name": "foo", ..., "shape": [1, d] }
    ]
}
```

As you can see from the shape, the input itself is already "stacked" (i.e. as a 2-dimensional `[1, d]` tensor). The adaptive batching code will then assume that the first dimension is the "batch dimension", and will just merge all the inputs into a `[n, d]` tensor, which then the HF pipeline rejects.

What I'm suggesting is that the inputs should be sent with shape `[d]` (i.e. as 1-dimensional tensors), and the adaptive batching code should then just merge all the inputs across that same dimension (i.e. as a `[n * d]` tensor). Does that make sense or am I missing something?

My main concern with the solution proposed in the PR is that it's quite ad-hoc to this pipeline, and we don't know if there's any other HF pipeline which may also accept a `array_inputs` kwarg and expects a 2-dimensional tensor there.

**Comment by saeid93** [comment]
Yes, you are right the `array_inputs` will be problematic unless we force the users to use it when they want to batch which might be impractical. So the pull request seems not a suitable solution. But the proposed solution isn't working either in the case of speech models. the `[d]` input will result in an `[n * d]` input to the server when there is `n` number of inputs. There are two problems with that:
1. The model will output the requests to the model as single output subject to the `[n * d]` input which in the case of speech models will be a long string of text that the corresponding ending point for each request is not distinguishable and it will result in having only one output request for the batched requests (which contains all the other requests resluts) and the rest of the requests will be empty (similar to the https://github.com/SeldonIO/MLServer/pull/738 that we discussed over [slack](https://seldondev.slack.com/archives/C03DQFTFXMX/p1663522460764359)).
2. In the case of speech models since the output is a dictionary even the single output is not returned and the `key` of the dictionary is returned.

## Steps to reproduce 1 and 2
using the following `model-settings.json` file:
```json
{
    "name": "transformer",
    "implementation": "mlserver_huggingface.HuggingFaceRuntime",
    "max_batch_size": 3,
    "max_batch_time": 1,
    "parameters": {
        "extra": {
            "task": "automatic-speech-recognition",
            "pretrained_model": "facebook/s2t-small-librispeech-asr",
            "batch_size": 3
        }
    }
}
```
and following minimal client for sending requests:
```python
import requests
from pprint import PrettyPrinter
from mlserver.types import InferenceResponse
pp = PrettyPrinter(indent=4)
from transformers import pipeline
from datasets import load_dataset
import threading

gateway_endpoint = "localhost:8080"
model = 'transformer'
endpoint = f"http://{gateway_endpoint}/v2/models/{model}/infer"

batch_test = 3
responses = []

ds = load_dataset(
    "hf-internal-testing/librispeech_asr_demo",
    "clean",
    split="validation")

input_data = ds[0]["audio"]["array"]

def send_requests():
    payload = {
        "inputs": [
            {
            "name": "inputs",
            "shape": [len(input_data)],
            "datatype": "FP32",
            "data": input_data.tolist(),
            "parameters": {
                "content_type": "np"
            }
            }
        ]
    }
    response = requests.post(endpoint, json=payload)
    responses.append(response)
    return response

thread_pool = []

for i in range(batch_test):
    t = threading.Thread(target=send_requests)
    t.start()
    thread_pool.append(t)

for t in thread_pool:
    t.join()

pp.pprint(inference_responses)
```
will result in:
```
[   InferenceResponse(model_name='transformer', model_version=None, id='775f6a5c-1904-4492-8710-8de220a08053', parameters=Parameters(content_type=None, headers=None), outputs=[ResponseOutput(name='output', shape=[93680], datatype='BYTES', parameters=Parameters(content_type='str', headers=None), data=TensorData(__root__=[]))]),
    InferenceResponse(model_name='transformer', model_version=None, id='da6d881c-2bb8-474f-982a-36f4450166fb', parameters=Parameters(content_type=None, headers=None), outputs=[ResponseOutput(name='output', shape=[93680], datatype='BYTES', parameters=Parameters(content_type='str', headers=None), data=TensorData(__root__=[]))]),
    InferenceResponse(model_name='transformer', model_version=None, id='e096f114-4c9a-4d54-a62d-2d6cc18ca35d', parameters=Parameters(content_type=None, headers=None), outputs=[ResponseOutput(name='output', shape=[93680], datatype='BYTES', parameters=Parameters(content_type='str', headers=None), data=TensorData(__root__=['"text"']))])]
```
As you see the 1st problem is evident as only one of the requests has an output and it doesn't output the full output (2nd) problem as it is only returning the key of the full output. Putting a logger at https://github.com/SeldonIO/MLServer/blob/e6b0bdb6aa7a281998b802324ec6dc724e12b8e7/runtimes/huggingface/mlserver_huggingface/runtime.py#L95 outputs the following which shows the output recieved from the model (that is the concatenation of the output as expected which has the 1st problem of no obvious boundray between them to unbatch):
```bash
ERROR:mlserver:prediction={'text': 'mister quilter is the apostle of the middle classes and we are glad to welcome his gospel mister quilter is the apostle of the middle classes and we are glad to welcome his gospel mister quilter is the apostle of the middle classes and we are glad to welcome his gospel'}
```

I spent some time to fix this through the batching code but couldn't understand the flow. If you could give me some clue about where exactly batching of numpy arrays (1 dimensional array) is happening I migth be able to update the pull request.

Thanks!

**Comment by adriangonz** [comment]
Thanks for investigating that approach and sharing those insights @saeid93. That's a really detailed report. Fully agree with you though, this is definitely not a trivial one....

Given that the workaround you've put together in #758 unblocks this one, I'd be keen to go ahead with that, even though we know may have some caveats. We can always merge and then go back to this discussion if any issues get reported.

BTW slightly related to this one, there's a separate PR from a community member which aims to improve how the different HF outputs get encoded as responses in the HF runtime. It would be massively appreciated to hear your thoughts on that one, as there's a chance some of those changes are relevant to this discussion. 

https://github.com/SeldonIO/MLServer/pull/692

**Comment by saeid93** [comment]
Great, Sure, I have a deeper look later this week on #692. Looks like the ultimate clean solution on the output side which I could also adapt on the input side of the model.

Cool, yes we can merge to solve it in the meantime but I'll be still happy to look for a more robust solution like #692. Please let me know if there is anything else needed for this workaround.

I think one possible solution might be to define a list type in the input and change the batching logic for that specific content_type, it would be more consistent with the rest of the api:
```
payload = {
    "inputs": [
        {
        "name": "inputs",
        "shape": [1, len(input_data)],
        "datatype": "FP32",
        "data": [input_data.tolist()],
        "parameters": {
            "content_type": "list"
        }
        }
    ]
}
```
I'll open a separate pull request for that (or any other cleaner solution) once I got a chance to work on it.

Thanks!

**Comment by pepesi** [comment]
I noticed this issue, then made some changes in this commit https://github.com/SeldonIO/MLServer/pull/692/commits/dec4a6127f98d6a9012af3b3d5393dbc1f1eca96
